refactor(platform-server): Add an ssr benchmark setup.

JeanMeche opened the pull request:
In order to investigate the performances of SSR, this commit introduces a benchmark suite which will measure several step of the rendering.

JeanMeche commented:
This PR does impact all the packages because of visibility changes, @josephperrott's approval should be enough for those.

AndrewKushnir commented on modules/ssr-benchmarks/src/app/app.component.ts:
```diff
@@ -0,0 +1,65 @@
+    this.finishTask = this.pendingTasks.add();
+  }
+
+  data$ = (this.getObservableFromTransferState() ?? from(testData())).pipe(
```
Quick question: why do we need to use observables here? I think it'd be great to keep this logic synchronous, so we don't measure this extra time in benchmarks.

JeanMeche replied:
`testData()` was asynchronous like in the original benchmark, https://github.com/eknkc/ssr-benchmark. 
I agree we could make it synchronous to simplifly the whole example.

AndrewKushnir commented on packages/platform-server/src/server.ts:
```diff
@@ -30,21 +30,27 @@ import {
 export const INTERNAL_SERVER_PLATFORM_PROVIDERS: StaticProvider[] = [
   {provide: DOCUMENT, useFactory: _document, deps: [Injector]},
   {provide: PLATFORM_ID, useValue: PLATFORM_SERVER_ID},
+  {provide: PLATFORM_ID, useValue: PLATFORM_SERVER_ID},
```
This looks like a duplicate of a previous line.

AndrewKushnir commented on packages/platform-server/src/server.ts:
```diff
@@ -57,8 +63,13 @@ export const INTERNAL_SERVER_PLATFORM_PROVIDERS: StaticProvider[] = [
 ];
 
 function initDominoAdapter() {
+  const enableDomino = inject(ENABLE_DOMINO, {optional: true}) ?? true;
```
Since we have it in 2 places and use a default value (of `true`), it'd be great to extract this logic into a separate function to make sure the default value remains the same in all places.

AndrewKushnir commented on packages/platform-server/src/server.ts:
```diff
@@ -89,11 +100,14 @@ export class ServerModule {}
 
 function _document(injector: Injector) {
   const config: PlatformConfig | null = injector.get(INITIAL_CONFIG, null);
+  const enableDomino = injector.get(ENABLE_DOMINO, true);
```
Looks like another location for the default value, we can also replace that with a utility function (see previous comment).

alan-agius4 replied:
Will `ENABLE_DOMINO` be used for other purposes in the future, such as disabling `DomEmulation`? While I understand the need for this, it seems like this functionality should be integrated into the Benchmark App and the Adapter / PlatformState be replaced via DI. 

I'm not sure if this is important in this context, but I think it's worth noting that Domino will still be very present.

This is because the first step in the server bundle, through `@angular/platform-server/init`, is to apply polyfills, patch `globalThis`, and replace/add the following symbols:

  - `CSSStyleDeclaration`
  - `CharacterData`
  - `Comment`
  - `DOMException`
  - `DOMImplementation`
  - `DOMTokenList`
  - `Document`
  - `DocumentFragment`
  - `DocumentType`
  - `Element`
  - `HTMLParser`
  - `NamedNodeMap`
  - `Node`
  - `NodeList`
  - `NodeFilter`
  - `ProcessingInstruction`
  - `Text`
  - `Window`
  - `Event`
  - `UIEvent`
  - `MouseEvent`
  - `CustomEvent`
  - `HTMLElement`
  - `HTMLUnknownElement`
  - `HTMLAnchorElement`
  - `HTMLAreaElement`
  - `HTMLBRElement`
  - `HTMLBaseElement`
  - `HTMLBodyElement`
  - `HTMLButtonElement`
  - `HTMLDListElement`
  - `HTMLDataElement`
  - `HTMLDataListElement`
  - `HTMLDetailsElement`
  - `HTMLDivElement`
  - `HTMLEmbedElement`
  - `HTMLFieldSetElement`
  - `HTMLFormElement`
  - `HTMLHRElement`
  - `HTMLHeadElement`
  - `HTMLHeadingElement`
  - `HTMLHtmlElement`
  - `HTMLIFrameElement`
  - `HTMLImageElement`
  - `HTMLInputElement`
  - `HTMLKeygenElement`
  - `HTMLLIElement`
  - `HTMLLabelElement`
  - `HTMLLegendElement`
  - `HTMLLinkElement`
  - `HTMLMapElement`
  - `HTMLMenuElement`
  - `HTMLMetaElement`
  - `HTMLMeterElement`
  - `HTMLModElement`
  - `HTMLOListElement`
  - `HTMLObjectElement`
  - `HTMLOptGroupElement`
  - `HTMLOptionElement`
  - `HTMLOutputElement`
  - `HTMLParagraphElement`
  - `HTMLParamElement`
  - `HTMLPreElement`
  - `HTMLProgressElement`
  - `HTMLQuoteElement`
  - `HTMLScriptElement`
  - `HTMLSelectElement`
  - `HTMLSpanElement`
  - `HTMLStyleElement`
  - `HTMLTableCaptionElement`
  - `HTMLTableCellElement`
  - `HTMLTableColElement`
  - `HTMLTableElement`
  - `HTMLTemplateElement`
  - `HTMLTableRowElement`
  - `HTMLTableSectionElement`
  - `HTMLTextAreaElement`
  - `HTMLTimeElement`
  - `HTMLTitleElement`
  - `HTMLUListElement`
  - `HTMLMediaElement`
  - `HTMLAudioElement`
  - `HTMLVideoElement`
  - `HTMLTableDataCellElement`
  - `HTMLTableHeaderCellElement`
  - `HTMLFrameSetElement`
  - `HTMLFrameElement`
  - `HTMLCanvasElement`
  - `HTMLDialogElement`
  - `HTMLMenuItemElement`
  - `HTMLSourceElement`
  - `HTMLTrackElement`
  - `HTMLFontElement`
  - `HTMLDirectoryElement`
  - `SVGElement`
  - `SVGSVGElement`
  - `KeyboardEvent`

alan-agius4 replied:
> ATM we disable Domino to run the benchmark in the browser (cf the README). 

But Domino is still around and patching the globals even with it being so called disabled. My point is that Domino, is not truly fully disabled.

JeanMeche replied:
In the benchmark I ran it was. As mentionned in the readme there is a manual step, I had to comment out the polyfill. 
It's not optimal I agree and we should look into removing that manual step. 
On the CLI side, we'll need to have an option/a way to drop this. 

https://github.com/angular/angular-cli/blob/99ee012b16a99b9a888262df8c95ae0df97922db/packages/angular/build/src/tools/esbuild/application-code-bundle.ts#L198-L205

alan-agius4 replied:
I am talking about Domino not `node:module`, the below will cause Domino to patch `globalThis` and replace all the above mentioned symbols.

https://github.com/angular/angular-cli/blob/99ee012b16a99b9a888262df8c95ae0df97922db/packages/angular/build/src/tools/esbuild/application-code-bundle.ts#L171

AndrewKushnir reviewed (commented):
Looks great, thanks for implementing this @JeanMeche 👍  Just left a few comments.

@alan-agius4 could you please take a look at this change when you get a chance?

Yberion commented:
Hello, is one of the next steps going to be testing against a regular CLI setup? (By that I mean, Domino, polyfill, Express/Fastify something "similar" to [this](https://github.com/platformatic/ssr-performance-showdown/pull/18))

Do you think it would be possible to use another dom renderer such as [happy-dom](https://github.com/capricorn86/happy-dom) in the futur? (It would require a new adapter)

alan-agius4 commented on modules/ssr-benchmarks/dist/modules/ssr-benchmarks/main.js:
```diff
@@ -0,0 +1,89 @@
+  render,
+  initData,
+  PERFORMANCE_MARK_PREFIX,
+} from '../../dist/out/darwin_arm64-fastbuild/bin/modules/ssr-benchmarks/build/server/main.server.mjs';
```
This path is unstable. The benchmarks are currently being executed outside the Bazel environment, which bypasses Bazel's linker and `require.resolve`. 

A more dependable solution would be to run the benchmarks directly through Bazel instead of using `tsx`. 

Alternatively, one could use [`write_file`](https://github.com/bazelbuild/bazel-skylib/blob/main/docs/write_file_doc.md) macro or a similar method to move the output of the architect rule to a directory relative to `modules/ssr-benchmarks/dist/modules/ssr-benchmarks`.

JeanMeche replied:
I agree this should be improve, it'll probably take me some time though as I'm pretty unfamiliar with this.

alan-agius4 replied:
https://github.com/angular/angular/pull/57647#discussion_r1754934053 is even a better alternative.

alan-agius4 commented on modules/ssr-benchmarks/BUILD.bazel:
```diff
@@ -0,0 +1,72 @@
+    ],
+)
+
+APPLICATION_DEPS = [
```
Can you clean the deps that are not being used?

alan-agius4 commented on packages/platform-server/src/utils.ts:
```diff
@@ -280,7 +290,8 @@ export async function renderApplication<T>(
-    return _render(platformRef, applicationRef);
+    const applicationRef = await runAndMeasurePerf('bootstrap', async () => await bootstrap());
+
+    return runAndMeasurePerf('_render', () => _render(platformRef, applicationRef));
```
@AndrewKushnir, I'm not sure if this is a concern, but we are introducing up to three nested layers of closures in a critical path. I'm unsure of the exact impact or how V8 optimizes this, but it may be worth considering.

AndrewKushnir replied:
I agree, let's brainstorm a different option. We can avoid nesting by adding start/stop calls, i.e.:
```
startMeasuring('_render');
_render(platformRef, applicationRef);
stopMeasuring('_render');
```
@JeanMeche @alan-agius4 WDYT? We can also followup offline to discuss details.

alan-agius4 replied:
Yeah, this does seem better.

JeanMeche replied:
Should we duplicate the labels or setup some sort of stack (from which we can push/pop the label) ?

AndrewKushnir replied:
> Should we duplicate the labels or setup some sort of stack (from which we can push/pop the label) ?

Keeping/repeating the label might be a bit easier - less code to implement.

alan-agius4 commented on packages/platform-server/src/platform_state.ts:
```diff
@@ -18,13 +19,14 @@ import {serializeDocument} from './domino_adapter';
  */
 @Injectable()
 export class PlatformState {
+  enableDomino = inject(ENABLE_DOMINO, {optional: true}) ?? true;
```
This seems like something we wouldn't want as part of the public API

JeanMeche commented:
@Yberion This is more intented to investigate Angular's internals.
The benchmarked app is build against the CLI architect (so with Domino, polyfills etc) but plugs directly to the `renderApplication` function so it dismisses the CommonEngine/Express and the Critical CSS inlining which is more of a CLI topic in itself.

AndrewKushnir commented on modules/ssr-benchmarks/README.md:
```diff
@@ -0,0 +1,49 @@
+│ 2       │ '  └ bootstrap'                      │ '0.1ms' │ '0.2ms'  │ '1.5%'     │ '6.0ms'  │
+│ 3       │ '  └ _render'                        │ '9.5ms' │ '11.7ms' │ '98.3%'    │ '32.3ms' │
+│ 4       │ '   └ whenStable'                    │ '4.8ms' │ '6.2ms'  │ '51.7%'    │ '17.5ms' │
+│ 5       │ '   └ prepareForHydration'           │ '1.5ms' │ '1.8ms'  │ '15.2%'    │ '6.4ms'  │
```
Yes, we'd need to look into that, it used to be 2-3%.

alan-agius4 commented on packages/platform-server/src/server.ts:
```diff
@@ -108,3 +120,8 @@ function _document(injector: Injector) {
   createPlatformFactory(platformCore, 'server', INTERNAL_SERVER_PLATFORM_PROVIDERS);
+
+function enableDomEmulation(injector: Injector): boolean {
+  const defaultValue = true;
+  return injector.get(ENABLE_DOM_EMULATION, defaultValue);
```
```suggestion
  return injector.get(ENABLE_DOM_EMULATION, true);
```

jkrems commented on modules/ssr-benchmarks/README.md:
```diff
@@ -0,0 +1,49 @@
+* Rendering is done against the local code of this repository (using the same trick as adev)
+
+* Individual function calls are measured with `runAndMeasurePerf` from the core package. 
+* If you add a new measure, make sur to add it also to the `levels` map for it to be represented correctly in the result
```
```suggestion
* If you add a new measure, make sure to add it also to the `levels` map for it to be represented correctly in the result
```

JeanMeche commented:
With the latest changes, the benchmark uses 2 bazel targets to run the benchmark with node and inside a browser runtime. 

Due to the CLI adding some polyfills, there are 2 dirty workarounds to disable DOM emulation and running server code inside a browser:  
1.  removing an import from `node:module` in `polyfills.server.mjs` (with `tail ...`)
2.  preventing from enabling the shims in `platform-server/init` via the global flag `NG_SKIP_DOMINO_SHIMS`.

alan-agius4 commented on modules/ssr-benchmarks/BUILD.bazel:
```diff
@@ -0,0 +1,25 @@
+        "yarn install --cache-folder ./.yarn_local_cache",
+        "yarn ng build --configuration production,browser",
+
+        ## This is a dirty workaround to drop an unnecessary import incompatible with running in a browser environment.
```
In the comment can you add which import is unnecessary here.

devversion commented on modules/ssr-benchmarks/BUILD.bazel:
```diff
@@ -17,9 +17,10 @@ ng_integration_test(
+        ## This is a dirty workaround to drop an import incompatible with running in a browser environment:
+        ## import { createRequire } from 'node:module';
+        ## globalThis['require'] ??= createRequire(import.meta.url);
         "tail -n +3 dist/ssr-benchmarks/server/polyfills.server.mjs > dist/ssr-benchmarks/server/polyfills.server.mjs",
```
Ideally make this a patch?

JeanMeche replied:
I failed at it at my first try because it `@angular/build` was a nested dependency. 
Patching the nested deps did the job.

JeanMeche commented on modules/ssr-benchmarks/src/main.server.ts:
```diff
@@ -0,0 +1,44 @@
+
+export {render, initData, PERFORMANCE_MARK_PREFIX};
+
+// Tooling expects a default export but we don't use/need it.
```
fyi @alan-agius4 , after updating to the v19 tooling, a default export was expected. Building would fail without it.

JeanMeche commented on modules/ssr-benchmarks/BUILD.bazel:
```diff
@@ -0,0 +1,22 @@
+load("//integration:index.bzl", "ng_integration_test")
+
+ng_integration_test(
```
Should we run this target as part of `test:ci` ? (By default it will).

JeanMeche commented:
PR is ready for re-review. 
The faulty linting job will fixed by an infra PR.

AndrewKushnir commented on modules/ssr-benchmarks/patches/@angular-devkit+build-angular++@angular+build+19.0.0-next.6+require.patch:
```diff
@@ -0,0 +1,42 @@
+diff --git a/node_modules/@angular-devkit/build-angular/node_modules/@angular/build/src/tools/esbuild/application-code-bundle.js b/node_modules/@angular-devkit/build-angular/node_modules/@angular/build/src/tools/esbuild/application-code-bundle.js
```
@JeanMeche sorry I've missed the discussion about this change, is it something that will be updated on the CLI side and we'll be able to remove the patch or the patch is needed permanently?

AndrewKushnir commented on modules/ssr-benchmarks/README.md:
```diff
@@ -0,0 +1,50 @@
+## Intro
```
nit (potentially for a followup PR): it'd be great to have some info on how to produce a flame chart as well.

AndrewKushnir commented on packages/platform-server/src/platform_state.ts:
```diff
@@ -18,13 +25,22 @@ import {serializeDocument} from './domino_adapter';
 @Injectable()
 export class PlatformState {
+  /* @internal */
+  _enableDomEmulation = inject(ENABLE_DOM_EMULATION, {optional: true}) ?? true;
```
Can we use the `enableDomEmulation` function here as well?

AndrewKushnir commented on packages/platform-server/src/platform_state.ts:
```diff
@@ -18,13 +25,22 @@ import {serializeDocument} from './domino_adapter';
-    return serializeDocument(this._doc);
+    const measuringLabel = 'renderToString';
+    startMeasuring(measuringLabel);
+    const rendered = this._enableDomEmulation
+      ? serializeDocument(this._doc)
+      : this._doc.documentElement.outerHTML;
```
I'd like to propose adding a quick comment here to explain that the `this._doc.documentElement.outerHTML` call is for browser-based tests.

AndrewKushnir replied:
We can also have a dev-mode check that we are in a browser environment.

AndrewKushnir reviewed (commented):
@JeanMeche thanks for addressing the feedback! I've left a few minor comment, but overall the changes look great 👍 

@alan-agius4 could you please also take a final look when you get a chance?

alan-agius4 commented on modules/ssr-benchmarks/src/app/test-data.ts:
```diff
@@ -0,0 +1,39 @@
+}
+
+export function initData(count: number) {
+  data = Array(count)
```
NIT: you can use `Array.from` here to prefill the array.

```js
data = Array.from({ length: count }, () => ({
  id: generateUUID(),
  name: generateUUID(),
}))
```

alan-agius4 commented on modules/ssr-benchmarks/src/app/test-data.ts:
```diff
@@ -0,0 +1,39 @@
+  return data;
+}
+
+// We can drop this and use crypto.randomUUID() once we move to node 20
```
Use `crypto.randomUUID` instead and configure the benchmark to use a node.js 20 toolchain.

```bazel
ng_integration_test(
   ....
   # TODO remove once the Node 20 is the default toolchain.
    toolchain = select({
        "@bazel_tools//src/conditions:linux_x86_64": "@node20_linux_amd64//:node_toolchain",
        "@bazel_tools//src/conditions:darwin": "@node20_darwin_amd64//:node_toolchain",
        "@bazel_tools//src/conditions:windows": "@node20_windows_amd64//:node_toolchain",
    }),
)
```

JeanMeche replied:
I couldn't determine why setting the toolchain to node 20 didn't change the actual version. The benchmark still ran with 18.20.

alan-agius4 commented on modules/ssr-benchmarks/tsconfig.app.json:
```diff
@@ -0,0 +1,10 @@
+/* To learn more about this file see: https://www.typescriptlang.org/docs/handbook/tsconfig-json.html. */
+{
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "outDir": "../../out-tsc/app",
+    "types": ["node"]
+  },
+  "files": ["src/main.ts", "src/main.server.ts", "main.ts"],
+  "include": ["src/**/*.d.ts", "../../node_modules/@types/dom-navigation/index.d.ts"]
```
Is `dom-navigation` needed? If so, reference types using the `types` options. 


```suggestion
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "../../out-tsc/app",
    "types": ["node", "dom-navigation"]
  },
  "files": ["src/main.ts", "src/main.server.ts", "main.ts"],
  "include": ["src/**/*.d.ts"]
```

alan-agius4 commented on modules/ssr-benchmarks/tsconfig.json:
```diff
@@ -0,0 +1,39 @@
+/* To learn more about this file see: https://www.typescriptlang.org/docs/handbook/tsconfig-json.html. */
```
This tsconfig should probable the "closer" to the one generated by the Angular CLI.

alan-agius4 commented on modules/ssr-benchmarks/main.ts:
```diff
@@ -0,0 +1,113 @@
+/**
```
Consider renaming this file to something more descriptive, such as `run-benchmark`. I stumbled upon this a few times during the review and confused it with `src/main.ts`.

alan-agius4 reviewed (commented):
Couple of NITs

JeanMeche commented on modules/ssr-benchmarks/main.ts:
```diff
@@ -0,0 +1,113 @@
+ */
+
+/* tslint:disable:no-console  */
+import {render, initData, PERFORMANCE_MARK_PREFIX} from './src/main.server';
```
I'll looking into this when the next pre-release is out!

alan-agius4 commented on modules/ssr-benchmarks/BUILD.bazel:
```diff
@@ -0,0 +1,41 @@
+)
+
+# This target is mostly intended for investigating via the devTools using the flamechart
+ng_integration_test(
```
Question why not use the normal server build? Chrome devtools can be used with a Node.js app.

JeanMeche committed:
refactor(platform-server): Add an ssr benchmark setup.

In order to investigate the performances of SSR, this commit introduces a benchmark suite which will measure several step of the rendering.

JeanMeche committed:
fixup! refactor(platform-server): Add an ssr benchmark setup.

josephperrott commented:
Caretaker Note: The licenses are okay, the tooling is having a problem parsing things licenses with an AND in them.

AndrewKushnir commented:
[Presubmit](http://test/OCL:681572822:BASE:681572147:1727901101178:f77e6b45).

JeanMeche committed:
fixup! refactor(platform-server): Add an ssr benchmark setup.

AndrewKushnir commented:
This PR was merged into the repository by commit 84b68969561b6f387ef8761218a89fb343fc4442.

The changes were merged into the following branches: main
